File: src/main/java/genesis/init/GenesisCreativeTabs.java
/*
 * This file is part of Genesis Mod, licensed under the MIT License (MIT).
 *
 * Copyright (c) 2017 Boethie
 * Copyright (c) 2017 contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package genesis.init;

import genesis.GenesisMod;
import net.minecraft.block.Block;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.function.Supplier;

public class GenesisCreativeTabs {

    public static final CreativeTabs BUILDING_BLOCKS = createTabWithBlockIcon("buildingBlocks", () -> GenesisBlocks.GRANITE);
    public static final CreativeTabs DECORATIONS = createTabWithBlockIcon("decorations", () -> GenesisBlocks.ARAUCARIOXYLON_SAPLING);
    public static final CreativeTabs MISC = createTabWithItemIcon("misc", () -> GenesisItems.RED_CLAY_BALL);
    public static final CreativeTabs MATERIALS = MISC;

    private static CreativeTabs createTabWithBlockIcon(final String label, final Supplier<Block> iconSupplier) {
        return createTabWithStackIcon(label, () -> new ItemStack(iconSupplier.get()));
    }

    private static CreativeTabs createTabWithItemIcon(final String label, final Supplier<Item> iconSupplier) {
        return createTabWithStackIcon(label, () -> new ItemStack(iconSupplier.get()));
    }

    private static CreativeTabs createTabWithStackIcon(final String label, final Supplier<ItemStack> iconSupplier) {
        return new CreativeTabs(GenesisMod.MOD_ID + "." + label) {
            @SideOnly(Side.CLIENT)
            @Override
            public ItemStack getTabIconItem() {
                return iconSupplier.get();
            }
        };
    }
}

File: src/main/java/genesis/init/GenesisCreativeTabs.kt
/*
 * This file is part of Genesis Mod, licensed under the MIT License (MIT).
 *
 * Copyright (c) 2017 Boethie
 * Copyright (c) 2017 contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package genesis.init

import genesis.GenesisMod
import net.minecraft.block.Block
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

object GenesisCreativeTabs {

    val BUILDING_BLOCKS = createTabWithBlockIcon("buildingBlocks", { GenesisBlocks.GRANITE })
    val DECORATIONS = createTabWithBlockIcon("decorations", { GenesisBlocks.ARAUCARIOXYLON_SAPLING })
    val MISC = createTabWithItemIcon("misc", { GenesisItems.RED_CLAY_BALL })
    val FOOD = createTabWithItemIcon("food", { GenesisItems.ZINGIBEROPSIS_RHIZOME })
    val MATERIALS = MISC

    private fun createTabWithBlockIcon(label: String, iconSupplier: () -> Block): CreativeTabs {
        return createTabWithStackIcon(label, { ItemStack(iconSupplier()) })
    }

    private fun createTabWithItemIcon(label: String, iconSupplier: () -> Item): CreativeTabs {
        return createTabWithStackIcon(label, { ItemStack(iconSupplier()) })
    }

    private fun createTabWithStackIcon(label: String, iconSupplier: () -> ItemStack): CreativeTabs {
        return object : CreativeTabs(GenesisMod.MOD_ID + "." + label) {
            @SideOnly(Side.CLIENT)
            override fun getTabIconItem(): ItemStack {
                return iconSupplier()
            }
        }
    }
}
